External/FEXCore/Source/Interface/HLE/Thunks/Thunks.cpp

```diff
@@ -11,47 +11,124 @@ tags: glue|thunks
 #include <FEXCore/Utils/LogManager.h>
 #include <FEXCore/IR/IR.h>
 #include <FEXCore/IR/IREmitter.h>
+#include "FEXCore/Utils/CompilerDefs.h"
 #include "Thunks.h"
 
+#include <cstdint>
 #include <dlfcn.h>
 
 #include <Interface/Context/Context.h>
 #include "FEXCore/Core/X86Enums.h"
 #include <malloc.h>
-#include <map>
+#include <mutex>
+#include <unordered_map>
+#include <unordered_set>
 #include <memory>
 #include <shared_mutex>
 #include <stdint.h>
 #include <string>
 #include <utility>
 
+
+#define HOST_TRAMPOLINE_ALLOC_STEP (16 * 1024)
+
 struct LoadlibArgs {
     const char *Name;
-    uintptr_t CallbackThunks;
 };
 
 static thread_local FEXCore::Core::InternalThreadState *Thread;
 
+struct TrampolineInstanceInfo {
+  uintptr_t HostPacker;
+  uintptr_t CallCallback;
+  uintptr_t GuestUnpacker;
+  uintptr_t GuestTarget;
+};
+
+struct GuestcallInfo {
+  uintptr_t GuestUnpacker;
+  uintptr_t GuestTarget;
+
+  [[nodiscard]] friend constexpr bool operator==(const GuestcallInfo&, const GuestcallInfo&) = default;
+};
+
+template <>
+struct std::hash<GuestcallInfo> {
+  size_t operator()(const GuestcallInfo& x) const noexcept {
+    return x.GuestTarget;
+  }
+};
+
+
+static __attribute__((aligned(16), naked, section("HostToGuestTrampolineTemplate"))) void HostToGuestTrampolineTemplate() {
+#if defined(_M_X86_64)
+  asm(
+    "lea 0f(%rip), %r11 \n"
+    "jmpq *0f(%rip) \n"
+    ".align 16 \n"
+    "0: \n"
+    ".quad 0, 0, 0, 0 \n"
+  );
+#elif defined(_M_ARM_64)
+  asm(
+    "adr x11, 0f \n"
+    "ldr x16, [x11] \n"
+    "br x16 \n"
+    ".align 16 \n"
+    "0: \n" // CallCallback
```

Comment:
```suggestion
    "0: \n" // TrampolineInstanceInfo
```

Reply:
Any reason for naming the label `0` and commenting it instead of just `TrampolineInstanceInfo:`?

Reply:
0 is a special label that can exist multiple times [0..9 are all special labels]. 0f means 0forward, ie the next occurance of 0.